loginchecker.php

```diff
@@ -0,0 +1,123 @@
+ *
+ * This can be run either as a web api, or on the CLI. When run on the
+ * CLI it conforms to the Nagios plugin standard.
+ *
+ * See also:
+ *  - http://nagios.sourceforge.net/docs/3_0/pluginapi.html
+ *  - https://nagios-plugins.org/doc/guidelines.html#PLUGOUTPUT
+ *
+ */
+
+defined('MOODLE_INTERNAL');
+
+$options = array(
+    'help' => false,
+    'critthresh' => 500,
+    'warnthresh' => 10,
+    'logtime' => 5,
+);
+
+if (isset($argv)) {
+    // If run from the CLI.
+    define('CLI_SCRIPT', true);
+    require_once(__DIR__ . '/../../../config.php');
+    require_once('nagios.php');
+    require_once($CFG->libdir . '/clilib.php');
+
+    list($options, $unrecognized) = cli_get_params($options,
+    array(
+        'h' => 'help',
+        )
+    );
+
+    $checktime = time() - ($options['logtime'] * 60);
+
+    if ($unrecognized) {
+        $unrecognized = implode("\n  ", $unrecognized);
+        cli_error(get_string('cliunknowoption', 'admin', $unrecognized));
+    }
+
+    if ($options['help']) {
+        print "Checks the moodle database for failed login attempts in a given time frame
+
+        loginchecker.php [options]
+
+        Options:
+        -h, --help            Print out this help
+            --critthresh=n    Threshold for number of failed logins to trigger a critical error (default 500 attempts)
+            --warnthresh=n    Threshold for number of failed logins to trigger a warning (default 10 attempts)
+            --logtime=n       Time in minutes to check back for a critical error (default 5 minutes prior)
+
+        Example:
+        \$sudo -u www-data /usr/bin/php admin/tool/heartbeat/loginchecker.php --logtime=60\n";
+
+        die;
+    }
+
+} else {
+    // If run from the web.
+    require_once(__DIR__ . '/../../../config.php');
+    require_once('nagios.php');
+
+    $options['critthresh'] = optional_param('critthresh', 500, PARAM_INT);
+    $options['warnthresh'] = optional_param('warnthresh', 10, PARAM_INT);
+    $options['logtime'] = optional_param('logtime', 5, PARAM_INT);
+    $checktime = time() - ($options['logtime'] * 60);
+
+    header("Content-Type: text/plain");
+
+    // Make sure varnish doesn't cache this. But it still might so go check it!
+    header('Pragma: no-cache');
+    header('Cache-Control: private, no-cache, no-store, max-age=0, must-revalidate, proxy-revalidate');
+    header('Expires: Tue, 04 Sep 2012 05:32:29 GMT');
+}
+
+global $DB;
+
+// Check that the database doesn't need to be upgraded.
+if (moodle_needs_upgrading()) {
+    send_critical("Moodle upgrade pending, login checker execution suspended");
+}
+
+$sqlstring = "SELECT count(*) AS logincount
+                FROM {logstore_standard_log}
+               WHERE target = 'user_login'
+                 AND timecreated > '$checktime'";
+
+$tablequery = $DB->get_record_sql($sqlstring);
+
+$faildlogcount = $tablequery->logincount;
+
+test_log_data($faildlogcount, $options);
+
+function test_log_data($count, $options) {
```

Comment:
Do this function have much value? The unit tests probably aren't super important here so I'd be ok to ditch them completely and the fold this function back into inline code